### ISSUE TITLE
Literal instance attribute recognized as string type

### ISSUE THREAD

**Issue by tea-de-kay** [microsoft/pyright#7659]
When assigning a parameter of type `Literal[...]` to an instance attribute in `__init__` via `self._attribute = param`, the instance attribute is recognized as being of type `str` in another method of that class.

Here is a minimal example:
```python
from typing import Literal

LiteralType = Literal["a", "b"]

class TestClass:
    def __init__(self, param: LiteralType) -> None:
        self._param1 = param
        self._param2: LiteralType = param

    def method1(self, param1: LiteralType, param2: LiteralType):
        pass

    def method2(self):
        self.method1(param1=self._param1, param2=self._param2)

```
In `method2` the first parameter gives a type error:
```
Argument of type "str" cannot be assigned to parameter "param1" of type "LiteralType" in function "method1"
  Type "str" cannot be assigned to type "LiteralType"
    "str" cannot be assigned to type "Literal['a']"
    "str" cannot be assigned to type "Literal['b']"
```
while the second parameter is fine.

Inside the `__init__` method, `param1` and `param2` are properly recognized as being of type `LiteralType`, so this
```python
...
def __init__(self, param: LiteralType) -> None:
        self._param1 = param
        self._param2: LiteralType = param
        self.method1(param1=self._param1, param2=self._param2)
...
```
does not give a type error.

I am using VS Code (1.88.0) with the Pylance Extension (v2024.4.1) and `basic` type checking mode.

**Comment by erictraut** [comment]
Pyright's behavior is correct and intentional here, so this isn't a bug.

If you don't provide a type declaration for an instance variable, its type is inferred based on values assigned to it within a class. Literal values are not retained as part of this inference. After all, if you were to assign `self._param1 = 1`, you probably wouldn't expect `_param1` to be inferred as `Literal[1]`; you'd expect it to be inferred as `int`. The same goes for all literal values. 

If you want an instance variable to limited to specific literal values, the correct way to do that is to provide an explicit type declaration for that instance variable, as you've done above for `_param2`.

**Comment by roganov** [comment]
I opened a similar issue https://github.com/microsoft/pyright/issues/7660 which I closed as a duplicate of this one.
Honestly, I find such behavior extremely surprising - how can the type of an instance variable be different in the `__init__` and another method?

**Comment by erictraut** [comment]
@roganov, it's important for you to differentiate between declared types and inferred types. These are fundamental concepts in static type checking, so if these are not clear in your mind, I recommend reviewing [this](https://microsoft.github.io/pyright/#/type-concepts) and [this](https://microsoft.github.io/pyright/#/type-inference?id=type-inference).
